src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java

```diff
@@ -2115,6 +2123,10 @@ public void visitParens(JCParens tree) {
     public void visitAssign(JCAssign tree) {
         Item l = genExpr(tree.lhs, tree.lhs.type);
         genExpr(tree.rhs, tree.lhs.type).load();
+        if (tree.lhs.type.isNonNullable()) {
+            code.emitop0(dup);
+            genNullCheck(tree.rhs);
```

Comment:
Should the cast always be generated even if we know that the RHS is non-nullable? (I'm not sure here)